SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java

```diff
@@ -119,14 +124,39 @@ public boolean apply(SingularityExecutorShellCommandDescriptor input) {
 
     final SingularityExecutorShellCommandDescriptor shellCommandDescriptor = matchingShellCommandDescriptor.get();
 
-    List<String> command = new ArrayList<>(shellCommandDescriptor.getCommand());
+    List<String> command = new ArrayList<>();
+
+    if (shellCommandDescriptor.isSwitchUser()) {
+      String switchUserCommand = String.format(executorConfiguration.getSwitchUserCommandFormat(), taskProcess.getTask().getExecutorData().getUser().or(executorConfiguration.getDefaultRunAsUser()));
+      command.addAll(Arrays.asList(switchUserCommand.split("\\s+")));
+    }
+
+    boolean isDocker = task.getTaskInfo().hasContainer() && task.getTaskInfo().getContainer().hasDocker();
+    if (isDocker) {
+      command.addAll(Arrays.asList("docker", "exec", String.format("%s%s", executorConfiguration.getDockerPrefix(), task.getTaskId())));
+    }
+
+    command.addAll(shellCommandDescriptor.getCommand());
 
     for (int i = 0; i < command.size(); i++) {
       if (command.get(i).equals(executorConfiguration.getShellCommandPidPlaceholder())) {
-        if (!taskProcess.getCurrentPid().isPresent()) {
-          throw new InvalidShellCommandException("No PID found");
+        int pid;
+        Path pidFilePath = MesosUtils.getTaskDirectoryPath(getTask().getTaskId()).resolve(executorConfiguration.getShellCommandPidFile());
+        if (Files.exists(pidFilePath)) {
+          try {
+            pid = Integer.parseInt(new Scanner(pidFilePath).useDelimiter("\\Z").next());
```

Comment:
`\Z` is the end of string anchor, used that so we are essentially just read the whole file as one string